versioned_docs/version-4.0.0-alpha.1/datastructures/custom.md

```diff
@@ -3,48 +3,124 @@ title: Custom Data Structures
 slug: /datastructures/custom-datastructure
 ---
 
-:::caution
-TODO
+The `ink_storage` crate provides useful utilities and data structures to organize and
+manipulate the contract's storage. However, contract authors should know that they can
+also create their own custom data structures.
 
-Beware, this page is no longer up to date for 4.0!
-:::
+## Using custom types on storage
+Any custom type wanting to be compatible with ink! storage must implement the 
+[`Storable`](https://docs.rs/ink_storage_traits/4.0.0-beta/ink_storage_traits/trait.Storable.html) 
+trait, so it can be SCALE
+[`encoded`](https://docs.rs/parity-scale-codec/3.2.2/parity_scale_codec/trait.Encode.html)
+and 
+[`decoded`](https://docs.rs/parity-scale-codec/3.2.2/parity_scale_codec/trait.Decode.html).
+Additionally, the traits 
+[`StorageLayout`](https://docs.rs/ink_storage/latest/ink_storage/traits/trait.StorageLayout.html)
+and [`TypeInfo`](https://docs.rs/scale-info/2.3.1/scale_info/trait.TypeInfo.html)
+are required as well. But don't worry, usually these traits can just be derived:
+
+```rust
+/// A custom type that we can use in our contract storage
+#[derive(scale::Decode, scale::Encode)]
+#[cfg_attr(
+    feature = "std",
+    derive(scale_info::TypeInfo, ink::storage::traits::StorageLayout)
+)]
+pub struct Inner {
+    value: bool,
+}
 
-While the `ink_storage` crate provides tons of useful utilities and data structures to organize and manipulate the contract's storage contract authors are not limited by its capabilities. By implementing the core `SpreadLayout`/`PackedLayout` traits (and the `StorageLayout` trait for supporting the metadata generated for the `.contract` bundle) users are able to define their very own custom storage data structures with their own set of requirement and features that work along the `ink_storage` data structures as long as they fulfill the mere requirements stated by those two traits.
+#[ink(storage)]
+pub struct MyContractStorage {
+    inner: Inner,
+}
+```
 
-A basic example of a custom struct is shown below:
+Even better: there is a macro `#[ink::storage_item]`, which derives all necessary traits for 
```

Comment:
Can you please add some notes about Packed/Non-Packed values because it is important to understand the context when this is useful to use?

Reply:
I'm not sure I understand, there's a whole chapter `Packed vs Non-Packed layout` in the storage layout docs explaining just that o?

Reply:
True, so just add a few sentences that this attribute is preferred to be used for non-packed values:

From [docs](https://paritytech.github.io/ink/ink_macro/attr.storage_item.html)

> If the type is non-packed it is best to rely on automatic storage key calculation via ink::storage_item

Reply:
Ah I see, thanks. Unfortunately the macro docs do not say why that it is the case, so I don't know what to write here. However I added this to be a link to the docs of the macro

Reply:
The `#[ink::storage_item]` macro is responsible for the storage key calculation for the non-`Packed` type with `AutoKey`. Without it, the key for non-`Packed` fields will be zero. I would say that the usage of this macro is mandatory if you don't plan to use `ManualKey`.

If you want to have your own implementation of the ink! storage traits for the non-`Packed` type, it still needs to use this macro only for the key calculation with `#[ink::storage_item(derive = false)]`.

If your type can be `Packed`, then you need to avoid usage of this macro and derive `scale::Encode` and `scale::Decode`.

Reply:
Thanks @xgreenx, I added it as a note